### ISSUE TITLE
Why use padding_mode='border'?

### ISSUE THREAD

**Issue by kwea123** [nianticlabs/monodepth2#20]
When warping the image, you use `padding_mode='border'`, which causes border effects (e.g. on the left and bottom for image 2->1)
https://github.com/nianticlabs/monodepth2/blob/ec10cf1a0cc06c8d9f7aad709650529e7284746d/trainer.py#L381-L384
![bordereffect](https://user-images.githubusercontent.com/11364490/59607236-e90d8f00-914d-11e9-80e1-2129efb28317.png)
In the code it seems that you compute the reconstruction loss on the whole image, including the border effect parts (if I'm wrong, correct me). Wouldn't it degrades the model's performance?

What I've seen others doing (e.g. struct2depth) is to use a mask, and only compute the loss on the masked (valid) area. This can be done using `padding_mode='zeros'` to generate the valid mask. E.g., black areas are not contributing to the loss.
![mask](https://user-images.githubusercontent.com/11364490/59607586-97b1cf80-914e-11e9-9b8d-c49793cc6582.png)

What's your opinion?

**Comment by mdfirman** [comment]
Thanks for the question! I'm going to leave this one for @mrharicot. Thanks!

**Comment by mrharicot** [comment]
Hi,
Without the `border` padding the network tends to predict zero disparities on the edge of the image in the stereo case.  Similar to this figure, taken from monodepth 1.
It also helped stabilize training in the first few iterations.
![image](https://user-images.githubusercontent.com/5363877/59808990-9352f700-92b3-11e9-816c-87c52d06f157.png)
We actually haven't evaluated whether it hurts or help depth prediction accuracy.

**Comment by kwea123** [comment]
Do you have any explanation why it predicts zero disparities on the edge or why it helps "stabilize training"?
For the zero disparities, the loss is computed on both 0->1 and 0->-1 so I think padding zeros on borders can still let the network learn from image 0->1?

**Comment by mdfirman** [comment]
Hi @kwea123 , I don't know exactly why these effects occur, but could take some guesses. It isn't too hard to change the network to have zero padding instead of 'border' padding, so it could be worth investigating this effect yourself. Do report back if you make any interesting findings!

Thanks,

**Comment by aj96** [comment]
@mrharicot @mdfirman When you used the zero padding as opposed to border padding, were you masking out the corresponding portion for the target frame? If you used zero padding when creating the warped frame, but you didn't mask out the corresponding portion for the target frame, this could lead to strange results. Ideally, you'd just apply the fly-out mask once to the view-synthesis loss, no?

**Comment by ClementPinard** [comment]
Hello,

be careful with zero padding, due to how interpolation is done, the pixels that are sampled right on the frontier will have a mix of gray (more specifically the color [0,0,0]) and the valid color. you need to explicitely filter out the out of bound pixels and not only padd with zeros. Here is a quick discussion about it : https://discuss.pytorch.org/t/spatial-transformer-networks-boundary-grid-interpolation-behaviour/8891

**Comment by mrharicot** [comment]
As @ClementPinard mentioned when using `padding_mode='zeros'` it causes a problem at the image edges as the network will essentially learn to do patch comparison and the only location which contains pure black edges is on the edge of the image which causes the network to match the edges of the images together resulting in zero disparity.
